### PR TITLE
CE-2834 Basic flow, Force open modal and submit type on publish

### PR DESCRIPTION
**Pull request by kamilkoterba**
Covered flows for editing template type in source editor.
1. Force modal on creating new tpl via URL
2. Force modal on publish if type is not defined

Synced together the rest of possible flows to use common functionality
3. Edit type on edit page via entrypoint for new page
4. Edit type on edit page via entrypoint for existing page
...

Provided different way of handling type save for editpage (on view page submit is asynch, on editpage it's stored in hidden field until publish)

Split editing on view page from editing on edit page and provied common modal control module

https://wikia-inc.atlassian.net/browse/CE-2834

@Wikia/community-engineering

### CODE REVIEW

File: extensions/wikia/EditPageLayout/js/plugins/TemplateClassificationEditorPlugin.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+				this.editor.on('save', this.proxy(this.forceType));
+			},
+
+			forceType: function () {
+				var	$typeField = $('#editform').find('input[name=templateClassificationType]'),
+					type;
+
+				if ($typeField.length > 0) {
+					type = $typeField.attr('value');
+					if (type !== '') {
```
Comment (lukaszkonieczny) [inline extensions/wikia/EditPageLayout/js/plugins/TemplateClassificationEditorPlugin.js]: I think `type` var is no needed here. You can use `$typeField.attr('value') !== ''` or `$typeField.attr('value').length`

File: extensions/wikia/TemplateClassification/TemplateClassification.hooks.php [hunk 2]
```diff
@@ -41,10 +96,10 @@ public function onBeforePageDisplay( \OutputPage $out, \Skin $skin ) {
 	 */
 	public function onPageHeaderPageTypePrepared( \PageHeaderController $pageHeaderController, \Title $title ) {
 		$user = $pageHeaderController->getContext()->getUser();
-		if ( ( new Permissions() )->shouldDisplayEntryPoint( $user, $title ) ) {
+		if ( ( new Permissions() )->shouldDisplayTypeLabel( $title ) ) {
```
Comment (lukaszkonieczny) [inline extensions/wikia/TemplateClassification/TemplateClassification.hooks.php]: I not sure is it permission. It's rather condition.

File: extensions/wikia/TemplateClassification/TemplateClassification.hooks.php [hunk 3]
```diff
@@ -27,9 +78,13 @@ public static function register() {
...
-		if ( ( new Permissions() )->shouldDisplayEntryPoint( $skin->getUser(), $out->getTitle() ) ) {
-			\Wikia::addAssetsToOutput( 'tempate_classification_js' );
-			\Wikia::addAssetsToOutput( 'tempate_classification_scss' );
+		$permissions = new Permissions();
+		if ( $permissions->shouldDisplayEntryPointOnView( $skin->getUser(), $out->getTitle() ) ) {
+			\Wikia::addAssetsToOutput( 'template_classification_in_view_js' );
+			\Wikia::addAssetsToOutput( 'template_classification_scss' );
+		} elseif ( $permissions->shouldDisplayEntryPointInEdit( $skin->getUser(), $out->getTitle() ) ) {
+			\Wikia::addAssetsToOutput( 'template_classification_in_edit_js' );
+			\Wikia::addAssetsToOutput( 'template_classification_scss' );
```
Comment (lukaszkonieczny) [inline extensions/wikia/TemplateClassification/TemplateClassification.hooks.php]: What do you think about using here just one method `shouldDisplayEntryPoint` and then in if statement check if it's 'edit' page.

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationInEdit.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+			return mw.config.get('wgArticleId') === 0;
+		}
+
+		function getType() {
+			var type;
+			if ($editFromHiddenTypeFiled.length === 0){
+				$editFromHiddenTypeFiled = $('#editform').find('[name=templateClassificationType]');
+			}
+			if ($editFromHiddenTypeFiled.length === 0) {
+				type = '';
```
Comment (lukaszkonieczny) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationInEdit.js]: You can init it in var declaration and check here if `length > 0`

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationInEdit.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+			} else {
+				type = $editFromHiddenTypeFiled.attr('value');
+			}
+			/* Return in format required by TemplateClassificationModal module */
+			return [{type:type}];
+		}
+
+		function storeTypeForSend(templateType) {
+			if ($editFromHiddenTypeFiled.length === 0) {
+				$editFromHiddenTypeFiled = $('<input>').attr({
```
Comment (lukaszkonieczny) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationInEdit.js]: I think this field should be already added in hook.

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationInView.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+		function init() {
+			templateClassificationModal.init(getType, storeTypeForSend);
+		}
+
+		function getType(articleId) {
+			if (selectedType) {
+				return [{type:selectedType}];
+			}
+
+			return nirvana.sendRequest({
```
Comment (lukaszkonieczny) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationInView.js]: Do we need this request while we already have this type generated on view page?

File: extensions/wikia/TemplateClassification/TemplateClassification.hooks.php [hunk 7]
```diff
@@ -15,7 +15,64 @@ public static function register() {
...
+	 * Save template type passed from article creation
+	 * template type is stored in templateClassificationType hidden field
+	 *
+	 * @param \WikiPage $wikiPage
+	 * @return bool
+	 */
+	public function onArticleInsertComplete( \WikiPage $wikiPage ) {
+		global $wgCityId;
+
+		( new \TemplateClassificationService() )->classifyTemplate(
```
Comment (lukaszkonieczny) [inline extensions/wikia/TemplateClassification/TemplateClassification.hooks.php:35]: `TemplateClassificationService` throws exception so we should catch them here in try/catch block

File: extensions/wikia/TemplateClassification/TemplateClassification.hooks.php [hunk 8]
```diff
@@ -15,7 +15,64 @@ public static function register() {
...
+
+	/**
+	 * Add hidden input to editform with template type
+	 * @param \EditPageLayout $editPage
+	 * @param \OutputPage $out
+	 */
+	public static function onEditPageShowEditFormFields( \EditPageLayout $editPage, \OutputPage $out ) {
+		global $wgCityId;
+		$articleId = $editPage->getTitle()->getArticleID();
+		$templateType = ( new \TemplateClassificationService() )->getType( $wgCityId, $articleId );
```
Comment (lukaszkonieczny) [inline extensions/wikia/TemplateClassification/TemplateClassification.hooks.php:75]: `TemplateClassificationService` throws exception so we should catch them here in try/catch block

File: extensions/wikia/TemplateClassification/TemplateClassificationView.php [hunk 9]
```diff
@@ -7,42 +7,63 @@ class View {
...
 	 * @return string
 	 */
-	public function renderTemplateType( $wikiId, $articleId, $user, $fallbackMsg = '' ) {
-		$templateType = lcfirst( ( new \TemplateClassificationService() )->getType( $wikiId, $articleId ) );
+	public function renderTemplateType( $wikiId, \Title $title, $user, $fallbackMsg = '', $templateTypeLabel = null ) {
+		if ( !$user->isLoggedIn() && !$this->isTemplateClassified( $title ) ) {
+			return $fallbackMsg;
+		}
+
+		$templateType = ( new \TemplateClassificationService() )->getType( $wikiId, $title->getArticleID() );
```
Comment (lukaszkonieczny) [inline extensions/wikia/TemplateClassification/TemplateClassificationView.php:21]: `TemplateClassificationService` throws exception so we should catch them here in try/catch block

... and all others :p

File: extensions/wikia/EditPageLayout/js/plugins/PageControls.js [hunk 10]
```diff
@@ -103,6 +103,13 @@
 
 		// handle "Save" button
 		onSave: function (event) {
+			event.preventDefault();
+
+			var eventOk = this.editor.fire('save');
```
Comment (lukaszkonieczny) [inline extensions/wikia/EditPageLayout/js/plugins/PageControls.js]: Do we need to create this variable?

File: extensions/wikia/TemplateClassification/TemplateClassificationView.php [hunk 11]
```diff
@@ -7,42 +7,63 @@ class View {
...
 	 * Renders an entry point on a template's edit page.
-	 * @param int $articleId
+	 * @param \Title $title
 	 * @param \User $user
 	 * @return string
 	 */
-	public function renderEditPageEntryPoint( $wikiId, $articleId, \User $user ) {
-		$templateType = $this->renderTemplateType( $wikiId, $articleId, $user );
+	public function renderEditPageEntryPoint( $wikiId, \Title $title, \User $user ) {
+		$templateType = $this->renderTemplateType( $wikiId, $title, $user, '', '' );
```
Comment (lukaszkonieczny) [inline extensions/wikia/TemplateClassification/TemplateClassificationView.php:67]: Do we need these two last params as they have default values in method

Comment (kamilkoterba) [inline extensions/wikia/TemplateClassification/TemplateClassificationView.php:67]: Note that second parameter is different than default that's why it's passed, but maybe we should rethink that if it's unclear?

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+		modalInstance.bind('done', function runSave() {
+			processSave(modalInstance);
+		});
+
+		/* Show the modal */
+		modalInstance.show();
+	}
+
+	function processSave(modalInstance) {
+		var templateType = $('#TemplateClassificationEditForm').serializeArray()[0].value;
```
Comment (lukaszkonieczny) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js]: Any reason for using `serialiazeArray` here?

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationInView.js [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+			return [{type: $typeLabel.data('type')}];
+		}
+
+		function storeTypeForSend() {
+			nirvana.sendRequest({
+				controller: 'TemplateClassificationApi',
+				method: 'classifyTemplate',
+				data: {
+					pageId: mw.config.get('wgArticleId'),
+					type: $('#TemplateClassificationEditForm').serializeArray()[0].value,
```
Comment (lukaszkonieczny) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationInView.js]: Any reason for using `serialiazeArray` here?

File: extensions/wikia/EditPageLayout/js/plugins/TemplateClassificationEditorPlugin.js [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+			init: function () {
+				this.editor.on('save', this.proxy(this.forceType));
+			},
+
+			forceType: function () {
+				var	$typeField = $('#editform').find('input[name=templateClassificationType]');
+
+				if ($typeField.length > 0) {
+					if ($typeField.attr('value') !== '') {
```
Comment (lukaszkonieczny) [inline extensions/wikia/EditPageLayout/js/plugins/TemplateClassificationEditorPlugin.js]: What about using just `val()` function?

File: extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+		}
+
+		if (!$classificationForm) {
+			classificationFormLoader = getTemplateClassificationEditForm;
+		}
+
+		// Fetch all data and open modal
+		$.when(
+			classificationFormLoader(),
+			typeGetter(mw.config.get('wgArticleId')),
```
Comment (lukaszkonieczny) [inline extensions/wikia/TemplateClassification/scripts/TemplateClassificationModal.js]: Do we need to pass `wgArticleId` here?

**Comment by lukaszkonieczny** [comment]
LGTM

File: extensions/wikia/TemplateClassification/TemplateClassification.hooks.php [hunk 16]
```diff
@@ -27,9 +85,18 @@ public static function register() {
...
-			\Wikia::addAssetsToOutput( 'tempate_classification_js' );
-			\Wikia::addAssetsToOutput( 'tempate_classification_scss' );
+		$title = $out->getTitle();
+		$permissions = new Permissions();
+		if ( $permissions->shouldDisplayEntryPoint( $skin->getUser(), $title )
+			&& $title->exists()
+		) {
+			\Wikia::addAssetsToOutput( 'template_classification_in_view_js' );
+			\Wikia::addAssetsToOutput( 'template_classification_scss' );
+		} elseif ( $permissions->shouldDisplayEntryPoint( $skin->getUser(), $title )
```
Comment (lukaszkonieczny) [inline extensions/wikia/TemplateClassification/TemplateClassification.hooks.php]: Do we need to check it twice `$permissions->shouldDisplayEntryPoint( $skin->getUser(), $title )` ?

**Comment by lukaszkonieczny** [comment]
One more comment. Feel free to move it to tesitng after it's fixed

**Comment by kamilkoterba** [comment]
Ready to merge after testing https://wikia-inc.atlassian.net/browse/CE-2834

**Pull request merged**